### PR TITLE
IGNITE-20879 Additional criterions for queries

### PR DESCRIPTION
**Pull request by nva**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:1]: Seems need to add example how user can use the new API. Right now I can't understand this

Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:1]: Added

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+            return this;
+        }
+
+        /**
+         * Set the given criteria.
+         *
+         * @param where where condition.
+         */
+        public SqlSerializer.Builder where(@Nullable Criteria where) {
```
Comment (ygerzhedovich) [inline modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:184]: What does mean `null` ?

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+import org.apache.ignite.table.criteria.CriteriaVisitor;
+import org.apache.ignite.table.criteria.Expression;
+import org.apache.ignite.table.criteria.Operator;
+import org.apache.ignite.table.criteria.Parameter;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Serializes {@link Criteria} into into SQL.
+ */
+public class SqlSerializer implements CriteriaVisitor<Void> {
```
Comment (ygerzhedovich) [inline modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:41]: what does mean Void in the context? Why it so?

File: modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        assertEquals("SELECT * FROM test WHERE NOT A IN (?, ?, ?)", ser.toString());
+        assertArrayEquals(new Object[] {"a", "b", "c"}, ser.getArguments());
+    }
+
+    @Test
+    void testAnd() {
+        var ser = new SqlSerializer.Builder()
+                .tableName("test")
+                .columns(Set.of("A", "B"))
+                .where(and(columnValue("a", nullValue()), not(columnValue("b", in("a", "b", "c")))))
```
Comment (ygerzhedovich) [inline modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java:80]: I've tried the following condition:
`.where(columnValue("a", equalTo(in(columnValue("a", nullValue()), columnValue("b", nullValue())))))`
Unfortunately no any error and I have result as `a = ?`
It look as incorrect

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+     * <pre>
+     *     columnValue(&quot;category&quot;, equalTo(&quot;toys&quot;))
+     * </pre>
+     *
+     * @param columnName Column name.
+     * @param condition Target condition.
+     * @return The created expression instance.
+     */
+    static Expression columnValue(String columnName, Condition condition) {
+        var oldElements = condition.getElements();
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:57]: using `var` for such caseses are forbidden

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+ */
+public interface Criteria {
+    /**
+     * Accept the visitor with the given context.
+     *
+     * @param <C> context type.
+     * @param v visitor.
+     * @param context context of visit.
+     */
+    <C> void accept(CriteriaVisitor<C> v, @Nullable C context);
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:42]: what purpose if the context? I don't see any real usage of it

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+ * </code></pre>
+ *
+ * @see CriteriaQuerySource
+ */
+public interface Criteria {
+    /**
+     * Accept the visitor with the given context.
+     *
+     * @param <C> context type.
+     * @param v visitor.
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:39]:
```suggestion
     * @param v Visitor.
```
Please check all javadocs

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java:1]: What reason have the class in public API?  Is it require for end user? The question also relate for part other classes in the package

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 6 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java]: Fixed

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 4 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java]: Fixed. Checked all other places

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java (hunk 1 above)
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:185]: null means return all entries in record view. Fixed javadoc.

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java (hunk 2 above)
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:42]: This means that this visitor doesn't require any context

File: modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java (hunk 3 above)
Comment (nva) [reply modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java:80]: This problem occurred due to https://github.com/apache/ignite-3/pull/2810#discussion_r1386303866.
Replaced `Object` with `Comparable` in argument and append another `equalsTo` method with byte[] as argument.

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java:1]: We can't remove as`Criteria#columnValue` use Column. 
It's like `Tuple` and `TupleImpl`. User can't create an instance of `TupleImpl` directly, but we store it in public api module.
So if we want to use `Criteria` as an entry point for building criteria, we need to put all these classes in a public api module. All these classes have a package visible constructor, so it shouldn't not confuse the user

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java (hunk 2 above)
Comment (ygerzhedovich) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:42]: can we add explanation to javadoc?

Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:44]: Improve javadoc

File: modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java (hunk 3 above)
Comment (ygerzhedovich) [reply modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java:80]: also need to have for all supported types

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 7]
```diff
@@ -17,9 +17,226 @@
...
+     *
+     * <p>For example:
+     * <pre>
+     *     columnValue("age", lessThanOrEqualTo(35))
+     * </pre>
+     *
+     * @param <T> Value type.
+     * @param value Target value.
+     */
+    static <T> Condition lessThanOrEqualTo(Comparable<T> value) {
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:196]: we should have tests for all methods for all types

File: modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Construct SQL query and arguments for prepare statement.
+     *
+     * @param tableName Table name.
+     * @param columns Table columns.
+     * @param criteria The predicate to filter entries or {@code null} to return all entries from the underlying table.
+     * @return SQL query and it's arguments.
+     */
+    static SqlSerializer createSqlSerializer(String tableName, ClientColumn[] columns, @Nullable Criteria criteria) {
```
Comment (ygerzhedovich) [inline modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java:60]: I can't find any tests related to the class

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 9]
```diff
@@ -17,9 +17,226 @@
...
+     * Creates a condition that test the examined object is equal to the specified {@code value}.
+     *
+     * <p>For example:
+     * <pre>
+     *     columnValue(&quot;category&quot;, equalTo(&quot;toys&quot;))
+     * </pre>
+     *
+     * @param value Target value.
+     */
+    static Condition equalTo(byte[] value) {
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:136]: should we have similar overloaded methods for another operations?

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.apache.ignite.table.criteria;
+
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a column reference for criteria query.
+ */
+public class Column implements Criteria {
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java:25]: seems all of this class should be final

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 11]
```diff
@@ -17,9 +17,226 @@
...
+    static <T> Condition equalTo(Comparable<T> value) {
+        return new Condition(Operator.EQ, new Parameter<>(value));
+    }
+
+    /**
+     * Creates a condition that test the examined object is equal to the specified {@code value}.
+     *
+     * <p>For example:
+     * <pre>
+     *     columnValue(&quot;category&quot;, equalTo(&quot;toys&quot;))
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:131]: example is inccorect.

File: modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                .build();
+
+        assertEquals("SELECT * FROM test WHERE (A IS NULL) AND (NOT B IN (?, ?, ?))", ser.toString());
+        assertArrayEquals(new Object[] {"a", "b", "c"}, ser.getArguments());
+    }
+
+    @Test
+    void testSqlInjection() {
```
Comment (ygerzhedovich) [inline modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java:88]: How it related to Sql injection?

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public <T> void visit(Expression expression, @Nullable Void context) {
+        Operator operator = expression.getOperator();
+        Criteria[] elements = expression.getElements();
+
+        String template;
+
+        if (operator == Operator.AND || operator == Operator.OR) {
```
Comment (ygerzhedovich) [inline modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:92]: we shoud have more tests for the code. As I see not covered cases with one parameter, for three parameters, no test for OR, and so on....

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+            template = IntStream.range(1, elements.length)
+                    .mapToObj(i -> String.format("{%d}", i))
+                    .collect(Collectors.joining(", ", prefix, ")"));
+        } else {
+            template = ELEMENT_TEMPLATES.get(operator);
+        }
+
+        int end = 0;
+        Matcher matcher = Pattern.compile("\\{(\\d+)\\}").matcher(template);
```
[the remark this answers is not in the document]
Comment (ygerzhedovich) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:109]: or better yet, insert parameters immediately when generating strings

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java
Comment (ygerzhedovich) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java:1]: seems you are right here

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 9 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:136]: I can't imagine any example for such a case.

File: modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java (hunk 8 above)
Comment (nva) [reply modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java]: That tests do you expect for an abstract class? This is the same as the `AbstractTableView` and there are no tests in it. If you speak about package-private visibility, I replaced it with protected.

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java (hunk 10 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Column.java]: Done

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 11 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java]: Fixed

File: modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java (hunk 12 above)
Comment (nva) [reply modules/table/src/test/java/org/apache/ignite/internal/table/criteria/SqlSerializerTest.java]: Column name can be used for sql inject if came from rest api, for example. Replaced with column name validation.

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java (hunk 13 above)
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java:90]: Extended

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+        Operator operator = expression.getOperator();
+        Criteria[] elements = expression.getElements();
+
+        String template;
+
+        if (operator == Operator.AND || operator == Operator.OR) {
+            String delimiter = operator == Operator.AND ? ") AND (" : ") OR (";
+
+            template = IntStream.range(0, elements.length)
+                    .mapToObj(i -> String.format("{%d}", i))
```
[the remark this answers is not in the document]
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java]: Removed

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 7 above)
Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:238]: Added

File: modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/table/criteria/SqlSerializer.java]: Move to the field

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java (hunk 9 above)
Comment (ygerzhedovich) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:149]: For example I would select rows for particular values of bynary columns, Why I can not do it?
like SELECT * FROM T WHERE bin_col in ('0xAA','0xBB')
the same for NOT IN

Comment (nva) [reply modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:149]: Added for in, not in

File: modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java [hunk 16]
```diff
@@ -290,6 +290,32 @@ static <T> Condition in(Comparable<T>... values) {
...
+     *
+     * <p>For example:
+     * <pre>
+     *     columnValue(&quot;password&quot;, in(&quot;MyPassword&quot;.getBytes(), &quot;MyOtherPassword&quot;.getBytes()))
+     * </pre>
+     *
+     * @param values The collection in which matching items must be found.
+     */
+    static Condition in(byte[]... values) {
+        if (values == null || values.length == 0) {
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/table/criteria/Criteria.java:304]: Why it can't be `null` or empty?

**Pull request merged**